### PR TITLE
Support for date/month date text segment filters

### PR DESCRIPTION
**Pull request by kuzmany**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | 
| New feature/enhancement? (use the a.x branch)      | ✅ 
| Deprecations?                          | 
| BC breaks? (use the c.x branch)        | 
| Automated tests included?              | not yet
| Related user documentation PR URL      | https://github.com/mautic/mautic-documentation/pull/266
| Related developer documentation PR URL | 
| Issue(s) addressed                     | 

#### Steps to test this PR:
This PR added nee date/month text string segment date filter

- day check current day to date/datetime field month
- month - check current month to date/datetime field month

1. Load up [this PR](https://mautibox.com)
2. Create custom date time field
3. Create contact with value
- if today is 19. 10. 2021
- then add to custom field 19.09.2020
Create another contact and add
- if today is 19. 10. 2021
- then add to custom field 07.19.2019
3. Create segment 1 with field to custom field equal = day
4. Create segment 2 with field to custom field equal = month
5. Rebuild segment
6. You should see first contact added to segment 1
7. You should see second contact added to segment 2

<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/10542"></a>

### CODE REVIEW

**Review by npracht — DISMISSED** [review]
Works perfectly from user perspective.
Needs code review.
Capture d’écran 2021-10-19 à 09 38 00
Capture d’écran 2021-10-19 à 09 38 08

File: app/bundles/LeadBundle/Segment/Decorator/Date/Day/DateDay.php [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+use Mautic\CoreBundle\Helper\DateTimeHelper;
+use Mautic\LeadBundle\Segment\ContactSegmentFilterCrate;
+
+class DateDay extends DateDayAbstract
+{
+    /**
+     * {@inheritdoc}
+     */
+    protected function modifyBaseDate(DateTimeHelper $dateTimeHelper)
```
Comment (escopecz) [inline app/bundles/LeadBundle/Segment/Decorator/Date/Day/DateDay.php]: Can this method be removed when it's empty?

File: app/bundles/LeadBundle/Segment/Decorator/Date/Day/DateDay.php [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    {
+    }
+
+    /**
+     * @return array|bool|float|string|null
```
Comment (escopecz) [inline app/bundles/LeadBundle/Segment/Decorator/Date/Day/DateDay.php]: Isn't it returning only string?

File: app/bundles/LeadBundle/Segment/Decorator/Date/Month/DateMonth.php [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    protected function modifyBaseDate(DateTimeHelper $dateTimeHelper)
+    {
+        $dateTimeHelper->setDateTime('midnight first day of this month', null);
+    }
+
+    /**
+     * @return array|bool|float|string|null
```
Comment (escopecz) [inline app/bundles/LeadBundle/Segment/Decorator/Date/Month/DateMonth.php]:
```suggestion
     * @return string
```

**Review by escopecz — CHANGES_REQUESTED** [review]
Please add tests

File: app/bundles/LeadBundle/Segment/Decorator/Date/Day/DateDay.php (hunk 1 above)
Comment (kuzmany) [reply app/bundles/LeadBundle/Segment/Decorator/Date/Day/DateDay.php]: Thanks for review. I clean up all reported code

**Comment by kuzmany** [comment]
@escopecz unit tests added

**Review by escopecz — CHANGES_REQUESTED** [review]
There are still uncovered lines that CodeCov is reporting. I cannot find a functional test as an example. We already have some in our fork. The test should:
1. create a date field
2. create several contacts with various date field values
3. create segment with the new filter
4. run the `mautic:segment:update` command
5. check the lead lists leads table if the contacts with matching date value are in the segment.

This way you ensure these new filters work now and won't be broken in a future update. In many cases a functional test is faster to write than cover all new methods with unit tests and it really gives you confidence that everything is working as intended.

**Comment by kuzmany** [comment]
@escopecz thanks. I did functional tests but it failed in another ListModelFunctionalTest
Should I delete my created segments after tests? I thought changes are rollback. Nothin help

Can you review it please?

**Comment by kuzmany** [comment]
Close at the moment. We will push it again once ready to M5

**Pull request closed without merging**